pandas/core/reshape/merge.py

```diff
@@ -105,22 +108,47 @@ def merge(
     indicator: bool = False,
     validate: str | None = None,
 ) -> DataFrame:
-    op = _MergeOperation(
-        left,
-        right,
-        how=how,
-        on=on,
-        left_on=left_on,
-        right_on=right_on,
-        left_index=left_index,
-        right_index=right_index,
-        sort=sort,
-        suffixes=suffixes,
-        copy=copy,
-        indicator=indicator,
-        validate=validate,
-    )
-    return op.get_result()
+    how = ("inner" if condition is None else "cross") if how is None else how
+
+    if condition is None:
+        res = _MergeOperation(
+            left,
+            right,
+            how=how,
+            on=on,
+            left_on=left_on,
+            right_on=right_on,
+            left_index=left_index,
+            right_index=right_index,
+            sort=sort,
+            suffixes=suffixes,
+            copy=copy,
+            indicator=indicator,
+            validate=validate,
+        ).get_result()
+    else:
+        if how != "cross":
+            raise MergeError(
+                'Must use `how="cross" | None` if `condition` is specified'
+            )
```

Comment:
@jreback proposing this restriction because otherwise we need to consider all the different SLAs surrounding result order and index values that come with different merge types, which will not be trivial to resolve with the lazy approach. Plus for all the other merge types, can just do a filter after the fact.